app/models/data_file.rb

```diff
@@ -103,4 +103,11 @@ def creator
     return unless create_audit
     create_audit.user
   end
+
+  def restore(child)
+    raise TrashbinParentException.new("#{kind} #{id} is deleted, and cannot restore its versions.::Restore #{kind} #{id}.") if is_deleted?
```

Comment:
Some of this error message formatting could go in the exception classes. Maybe pass them `kind` and `id`, or an object?

Reply:
This is nice, but I think it can wait until the hardening phase in a couple of weeks. It isnt something that prevents code from moving forward.